lib/utils/parse/parse_test.go

```diff
@@ -155,99 +165,83 @@ func TestInterpolate(t *testing.T) {
+			in:     "{{external.foo}}",
+			traits: map[string][]string{"foo": {"a", "b"}, "bar": {"c"}},
 			res:    result{values: []string{"a", "b"}},
 		},
 		{
 			title:  "mapped traits with email.local",
-			in:     Expression{variable: "foo", transform: emailLocalTransformer{}},
-			traits: map[string][]string{"foo": []string{"Alice <alice@example.com>", "bob@example.com"}, "bar": []string{"c"}},
+			in:     "{{email.local(external.foo)}}",
+			traits: map[string][]string{"foo": {"Alice <alice@example.com>", "bob@example.com"}, "bar": {"c"}},
 			res:    result{values: []string{"alice", "bob"}},
 		},
 		{
 			title:  "missed traits",
-			in:     Expression{variable: "baz"},
-			traits: map[string][]string{"foo": []string{"a", "b"}, "bar": []string{"c"}},
+			in:     "{{external.baz}}",
+			traits: map[string][]string{"foo": {"a", "b"}, "bar": {"c"}},
 			res:    result{err: trace.NotFound("not found"), values: []string{}},
 		},
 		{
 			title:  "traits with prefix and suffix",
-			in:     Expression{prefix: "IAM#", variable: "foo", suffix: ";"},
-			traits: map[string][]string{"foo": []string{"a", "b"}, "bar": []string{"c"}},
+			in:     "IAM#{{external.foo}};",
+			traits: map[string][]string{"foo": {"a", "b"}, "bar": {"c"}},
 			res:    result{values: []string{"IAM#a;", "IAM#b;"}},
 		},
 		{
 			title:  "error in mapping traits",
-			in:     Expression{variable: "foo", transform: emailLocalTransformer{}},
-			traits: map[string][]string{"foo": []string{"Alice <alice"}},
+			in:     "{{email.local(external.foo)}}",
+			traits: map[string][]string{"foo": {"Alice <alice"}},
 			res:    result{err: trace.BadParameter("")},
 		},
 		{
 			title:  "literal expression",
-			in:     Expression{namespace: LiteralNamespace, variable: "foo"},
-			traits: map[string][]string{"foo": []string{"a", "b"}, "bar": []string{"c"}},
+			in:     "foo",
+			traits: map[string][]string{"foo": {"a", "b"}, "bar": {"c"}},
 			res:    result{values: []string{"foo"}},
 		},
 		{
-			title: "regexp replacement with numeric match",
-			in: Expression{
-				variable: "foo",
-				transform: regexpReplaceTransformer{
-					re:          regexp.MustCompile("bar-(.*)"),
-					replacement: "$1",
-				},
-			},
-			traits: map[string][]string{"foo": []string{"bar-baz"}},
+			title:  "regexp replacement with numeric match",
+			in:     `{{regexp.replace(internal.foo, "bar-(.*)", "$1")}}`,
+			traits: map[string][]string{"foo": {"bar-baz"}},
 			res:    result{values: []string{"baz"}},
 		},
-		{
-			title: "regexp replacement with named match",
-			in: Expression{
-				variable: "foo",
-				transform: regexpReplaceTransformer{
-					re:          regexp.MustCompile("bar-(?P<suffix>.*)"),
-					replacement: "${suffix}",
-				},
+		/*
+			TODO: the following test case fails due to https://github.com/gravitational/teleport/issues/16797
+			{
+				title:  "regexp replacement with named match",
+				in:     `{{regexp.replace(internal.foo, "bar-(?P<suffix>.*)", "${suffix}")}}`,
```

Comment:
would it work with `$suffix` instead of `${suffix}`?

Reply:
It does, you're right: 0a7da411e7913dc19e64fa7e9d626ad71b6829b6

#16797 is about something else: `$foo_` is interpreted as variable `foo_` instead of variable `foo` + `"_"`, and it's not possible to work around it since we disallow `${foo}`.